pkg/tc/qdisc.go

```diff
@@ -0,0 +1,343 @@
+			i--
+		case "clsact":
+			ret.obj.Kind = "clsact"
+
+			if ret.parent != nil {
+				return nil, ErrInvalidArg
+			}
+
+			indirectPar := tc.HandleIngress // is the same as clsact handle
+			ret.parent = &indirectPar
+
+			indirectHan := core.BuildHandle(tc.HandleIngress, 0)
+			ret.handle = &indirectHan
+			i--
+		case "parent":
+			qdiscID, err := strconv.ParseUint(val, 16, 32)
+			if err != nil {
+				return ret, err
+			}
+			if qdiscID >= 0x7FFFFFFF {
+				return nil, ErrOutOfBounds
+			}
+			indirect := uint32(qdiscID)
+			ret.parent = &indirect
+		case "estimator":
+			return nil, ErrNotImplemented
+		case "stab":
+			return nil, ErrNotImplemented
+		case "ingress_block":
+			return nil, ErrNotImplemented
+		case "egress_block":
+			return nil, ErrNotImplemented
+		case "help":
+			PrintQdiscHelp(stdout)
+		default:
+			fmt.Println(args[i:])
+			// Resolve Qdisc and parameters
+			var qdiscParseFn func([]string, io.Writer) (*tc.Object, error)
+			if qdiscParseFn = supportetQdisc(args[i]); qdiscParseFn == nil {
+				return nil, fmt.Errorf("%w: invalid qdisc: %s", ErrInvalidArg, args[i])
+			}
+			var err error
+			ret.obj, err = qdiscParseFn(args[i+1:], stdout)
+			if err != nil {
+				return nil, err
+			}
+			ret.kind = &ret.obj.Kind
+
+			return ret, nil
+		}
+	}
+
+	return ret, nil
+}
+
+const (
+	QdiscHelp = `Usage:
+	tc qdisc [ add | del | replace | change | show ] dev STRING
+   		[ handle QHANDLE ] [ root | ingress | clsact | parent CLASSID ]
+   		[ estimator INTERVAL TIME_CONSTANT ]
+  		[ stab [ help | STAB_OPTIONS] ]
+  		[ ingress_block BLOCK_INDEX ] [ egress_block BLOCK_INDEX ]
+  		[ [ QDISC_KIND ] [ help | OPTIONS ] ]
+
+	tc qdisc { show | list } [ dev STRING ] [ QDISC_ID ] [ invisible ]
+
+	Where:
+	QDISC_KIND := { [p|b]fifo | tbf | prio | red | etc. }
+	OPTIONS := ... try tc qdisc add <desired QDISC_KIND> help
+	STAB_OPTIONS := ... try tc qdisc add stab help
+	QDISC_ID := { root | ingress | handle QHANDLE | parent CLASSID }`
+)
+
+func PrintQdiscHelp(stdout io.Writer) {
+	fmt.Fprintf(stdout, "%s", QdiscHelp)
+}
+
+func (t *Trafficctl) ShowQdisc(args *QArgs, stdout io.Writer) error {
+	qdiscs, err := t.Tc.Qdisc().Get()
+	if err != nil {
+		return err
+	}
+
+	for _, qdisc := range qdiscs {
+		iface, err := net.InterfaceByIndex(int(qdisc.Ifindex))
+		if err != nil {
+			return err
+		}
+
+		if args.dev != "" {
+			if args.dev == iface.Name {
+				fmt.Fprintf(stdout, "%20s\t%s\n", iface.Name, qdisc.Kind)
+			}
+			continue
+		}
+		fmt.Fprintf(stdout, "%20s\t%s\n", iface.Name, qdisc.Kind)
+	}
+	return nil
+}
+
+func (t *Trafficctl) AddQdisc(args *QArgs, stdout io.Writer) error {
+	iface, err := net.InterfaceByName(args.dev)
+	if err != nil {
+		return err
+	}
+
+	msg := tc.Msg{
+		Family:  unix.AF_UNSPEC,
+		Ifindex: uint32(iface.Index),
+		Parent:  *args.parent,
+		Handle:  *args.handle,
+	}
+
+	obj := &tc.Object{
+		Msg:       msg,
+		Attribute: args.obj.Attribute,
+	}
+
+	fmt.Printf("%v\n", obj.Attribute.Htb.Init.Defcls)
```

Comment:
The pretty printing gives me headaches, because to implement such thing properly, I have to fall back to reflections for basically everything. There're no string functions in the library for the structures. This shall not be the final printing, but I'm unsure how to solve it yet.